lib/Sema/CodeSynthesis.cpp

```diff
@@ -864,51 +864,69 @@ static bool hasUserDefinedDesignatedInit(Evaluator &eval,
                            false);
 }
 
+static bool canInheritDesignatedInits(Evaluator &eval, ClassDecl *decl) {
+  // We can only inherit designated initializers if the user hasn't defined
+  // a designated init of their own, and all the stored properties have initial
+  // values.
+  return !hasUserDefinedDesignatedInit(eval, decl) &&
+         areAllStoredPropertiesDefaultInitializable(eval, decl);
+}
+
+static void collectNonOveriddenSuperclassInits(
+    ClassDecl *subclass, SmallVectorImpl<ConstructorDecl *> &results) {
+  auto superclassTy = subclass->getSuperclass();
+  assert(superclassTy);
+
+  // Record all of the initializers the subclass has overriden, excluding stub
+  // overrides, which we don't want to consider as viable delegates for
+  // convenience inits.
+  llvm::SmallPtrSet<ConstructorDecl *, 4> overriddenInits;
+  for (auto member : subclass->getMembers())
+    if (auto ctor = dyn_cast<ConstructorDecl>(member))
+      if (!ctor->hasStubImplementation())
+        if (auto overridden = ctor->getOverriddenDecl())
+          overriddenInits.insert(overridden);
+
+  auto superclassCtors = TypeChecker::lookupConstructors(
+      subclass, superclassTy, NameLookupFlags::IgnoreAccessControl);
+
+  for (auto memberResult : superclassCtors) {
+    auto superclassCtor = cast<ConstructorDecl>(memberResult.getValueDecl());
+
+    // Skip invalid superclass initializers.
+    if (superclassCtor->isInvalid())
+      continue;
+
+    // Skip unavailable superclass initializers.
+    if (AvailableAttr::isUnavailable(superclassCtor))
+      continue;
+
+    if (!overriddenInits.count(superclassCtor))
+      results.push_back(superclassCtor);
+  }
+}
+
 /// For a class with a superclass, automatically define overrides
 /// for all of the superclass's designated initializers.
 static void addImplicitInheritedConstructorsToClass(ClassDecl *decl) {
   // Bail out if we're validating one of our constructors already;
   // we'll revisit the issue later.
-  if (!decl->hasClangNode()) {
-    for (auto member : decl->getMembers()) {
-      if (auto ctor = dyn_cast<ConstructorDecl>(member)) {
-        if (ctor->isRecursiveValidation())
-          return;
-      }
+  for (auto member : decl->getMembers()) {
+    if (auto ctor = dyn_cast<ConstructorDecl>(member)) {
+      if (ctor->isRecursiveValidation())
```

Comment:
@slavapestov Is there a reason we need any of this as a preflight?  It seems bogus to me.

Reply:
I did try removing it, and it passed the validation test suite, I just wasn't sure whether or not it might cause us to start emitting circularity diagnostics for code that we previously silently accepted, so thought I'd leave it for a future PR.

I'd be more than happy to remove it now if you and Slava don't think it'd be a problem :)